apps/fault/src/main.c

```diff
@@ -80,6 +82,213 @@ static inline seL4_Word fault_handler_start(seL4_CPtr ep, seL4_CPtr done_ep, seL
+
+static void __attribute__((noinline)) read_fault(void)
+{
+    int *x = (int *)BAD_VADDR;
+    /* Do a read fault. */
+#if defined(CONFIG_ARCH_AARCH32)
+    asm volatile(
+        "read_fault_address:\n\t"
+        "ldr r0, [%[addrreg]]\n\t"
+        "read_fault_restart_address:\n\t"
+        :
+        : [addrreg] "r"(x)
+        : "r0"
+    );
+#elif defined(CONFIG_ARCH_AARCH64)
+    asm volatile(
+        "read_fault_address:\n\t"
+        "ldr x0, [%[addrreg]]\n\t"
+        "read_fault_restart_address:\n\t"
+        :
+        : [addrreg] "r"(x)
+        : "x0"
+    );
+#elif defined(CONFIG_ARCH_RISCV)
+    asm volatile(
+        "read_fault_address:\n\t"
+        LOAD_S " a0, 0(%[addrreg])\n\t"
+        "read_fault_restart_address:\n\t"
+        :
+        : [addrreg] "r"(x)
+        : "a0"
+    );
+#elif defined(CONFIG_ARCH_X86)
+    asm volatile(
+        "read_fault_address:\n\t"
+        "mov (%[addrreg]), %%eax\n\t"
+        "read_fault_restart_address:\n\t"
+        :
+        : [addrreg] "r"(x)
+        : "eax"
+    );
+#else
+#error "Unknown architecture."
+#endif
+}
+
+static void __attribute__((noinline))
+set_good_magic_and_set_pc(seL4_CPtr tcb, seL4_Word new_pc)
+{
+    /* Set their register to GOOD_MAGIC and set PC past fault. */
+    int error;
+    seL4_UserContext ctx;
+    error = seL4_TCB_ReadRegisters(tcb,
+                                   false,
+                                   0,
+                                   sizeof(ctx) / sizeof(seL4_Word),
+                                   &ctx);
+    assert(!error);
+#if defined(CONFIG_ARCH_AARCH32)
+    assert(ctx.r0 == BAD_MAGIC);
+    ctx.pc = new_pc;
+#elif defined(CONFIG_ARCH_AARCH64)
+    ctx.pc = new_pc;
+#elif defined(CONFIG_ARCH_RISCV)
+    ctx.pc = new_pc;
+#elif defined(CONFIG_ARCH_X86_64)
+    ctx.rip = new_pc;
+#elif defined(CONFIG_ARCH_IA32)
+    ctx.eip = new_pc;
+#else
+#error "Unknown architecture."
+#endif
+    error = seL4_TCB_WriteRegisters(tcb,
+                                    false,
+                                    0,
+                                    sizeof(ctx) / sizeof(seL4_Word),
+                                    &ctx);
+    assert(!error);
+}
+
+static void measure_vm_fault_fn(int argc, char **argv)
+{
+    assert(argc == N_FAULTER_ARGS);
+    volatile ccnt_t *start = (volatile ccnt_t *) atol(argv[0]);
+    seL4_CPtr done_ep = atol(argv[2]);
+
+    for (int i = 0; i < N_RUNS + 1; i++) {
+        SEL4BENCH_READ_CCNT(*start);
+        read_fault();
+    }
+
+    seL4_Send(done_ep, seL4_MessageInfo_new(0, 0, 0, 0));
+}
+
+static void measure_vm_fault_handler_fn(int argc, char **argv)
+{
+    seL4_CPtr ep, done_ep, reply, tcb;
+    volatile ccnt_t *start;
+    ccnt_t end;
+    fault_results_t *results;
+    seL4_Word badge = 0;
+    seL4_MessageInfo_t tag = seL4_MessageInfo_new(0, 0, 0, 0);
+
+    parse_handler_args(argc, argv, &ep, &start, &results, &done_ep, &reply, &tcb);
+
+    seL4_Word junk;
```

Comment:
this variable is not used anywhere?

Reply:
Yes, I think this is an artifact of another approach I was trying to take earlier. Will remove later once there are more edits to be done.